File: src/mission_pkg/src/mission_switcher.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/bool.hpp"
#include <cstdlib>
#include <string>

class MissionSwitcher : public rclcpp::Node
{
public:
    MissionSwitcher() : Node("mission_switcher"), switched_(false)
    {
        sub_ = this->create_subscription<std_msgs::msg::Bool>(
            "/switch_to_navigation", 10,
            std::bind(&MissionSwitcher::callback, this, std::placeholders::_1)
        );

        RCLCPP_INFO(this->get_logger(), "Waiting for /switch_to_navigation...");
    }

private:
    void callback(const std_msgs::msg::Bool::SharedPtr msg)
    {
        if (msg->data && !switched_)
        {
            RCLCPP_INFO(this->get_logger(), "Switching from exploration to navigation...");

            // Stop explorer node (namespace is /explorer)
            std::system("ros2 node kill /explorer/explorer_node");

            // Launch far_planner node (adjust executable name if needed)
            std::system("ros2 run far_planner far_planner_node &");

            // Publish goal
            std::string goal_cmd =
                "ros2 topic pub /goal_point geometry_msgs/msg/PointStamped "
                "'{header: {frame_id: \"map\"}, point: {x: -3.0, y: 2.0, z: 0.0}}' --once";

            std::system(goal_cmd.c_str());

            switched_ = true;
        }
    }

    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr sub_;
    bool switched_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MissionSwitcher>());
    rclcpp::shutdown();
    return 0;
}

File: src/mission_pkg/scripts/mission_switcher.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from std_msgs.msg import Bool
import subprocess

class MissionSwitcher(Node):
    def __init__(self):
        super().__init__('mission_switcher')
        self.subscription = self.create_subscription(
            Bool,
            '/switch_to_navigation',
            self.switch_callback,
            10
        )
        self.exploration_running = True
        self.get_logger().info("Waiting for /switch_to_navigation...")

    def switch_callback(self, msg):
        if msg.data and self.exploration_running:
            self.get_logger().info("Switch triggered! Stopping exploration and starting navigation.")

            # Kill explorer (if launched from separate launch)
            subprocess.run(['pkill', '-f', 'explore_robotx.launch'])

            # Start navigation
            subprocess.Popen(['ros2', 'launch', 'smb_bringup', 'smb_sim_navigation.launch.py'])

            # Optionally send a goal
            goal = """header:
                    stamp:
                        sec: 0
                        nanosec: 0
                    frame_id: ''
                    point:
                    x: -3.0
                    y: 2.0
                    z: 0.0
                    """
            subprocess.run([
                'ros2', 'topic', 'pub', '/goal_point', 'geometry_msgs/msg/PointStamped', goal, '--once'
            ])

            self.exploration_running = False

